spec.md

```diff
@@ -1586,12 +1760,192 @@ en <= Z_valid
+Every sink-flow probe must be the target of exactly one of these statements.
+
+These statements are detailed below.
+
+### Export
+
+The export statement takes a reference expression, which must be a probe expression,
+and exports the probe reference to into a sink-flow static reference target.
+
+```firrtl
+module Refs:
+  input clock:  Clock
+  output a : Probe<{x: UInt, y: UInt}> ; read-only ref. to wire 'p'
+  output b : RWProbe<UInt> ; force-able ref. to node 'q', inferred width.
+  output c : Probe<UInt<1>> ; read-only ref. to register 'r'
+  output d : RWProbe<UInt<2>>[4] ; vector of ref.'s to memory data in 'm'
+  output e : Probe<Clock> ; ref. to input clock port
+
+  wire p : {x: UInt, flip y : UInt}
+  export probe(p) as a ; probe is passive
+  node q = UInt<1>(0)
+  export rwprobe(q) as b
+  reg r: UInt, clock
+  export probe(r) as c
+  mem m:
+    data-type => UInt<5>
+    depth => 4
+    ; ...
+    read-under-write => undefined
+
+  export rwprobe(m) as d
+  export probe(clock) as e
+```
+
+Exporting to a field within a bundle or other statically known sub-element of
+an aggregate is allowed, for example:
+```firrtl
+module Foo:
+  input x : UInt
+  output y : {x: UInt, p: Probe<UInt>}
+  output z : Probe<UInt>[2]
+
+  wire p : UInt
+  p <= x
+  y.x <= p
+
+  export probe(p) as y.p
+  export probe(p) as z[0]
+  export probe(p) as z[1]
+```
+
+`RWProbe`{.firrtl} references to ports are not allowed on public-facing modules.
+
+#### Probes and Passive Types
+
+While `Probe`{.firrtl} inner types are passive, the type of the exported expression
+is not required to be:
+
+```firrtl
+module Foo :
+  input x : {a: UInt, flip b: UInt}
+  output y : {a: UInt, flip b: UInt}
+  output xp : Probe<{a: UInt, b: UInt}> ; passive
+
+  wire p : {a: UInt, flip b: UInt} ; p is not passive
+  export probe(p) as xp
+  p <= x
+  y <= p
+```
+
+#### Exporting References to Nested Declarations
+
+Nested declarations (see [@sec:nested-declarations]) may be exported:
+
+```firrtl
+module RefProducer :
+  input a : UInt
+  input en : UInt<1>
+  output thereg : Probe<UInt>
+
+  when en :
+    reg myreg : UInt, clk
+    myreg <= a
+    export probe(myreg) as thereg
+```
+
+### Forward
+
+The forward statement is similar to export but forwards an existing reference
+to the specified target.
+
+This can be used to pass a child module's reference further up the hierarchy:
+
+```firrtl
+module Foo :
+  output p : Probe<UInt>
+  ; ...
+
+module Forward :
+  output p : Probe<UInt>
+
+  inst f of Foo
+  forward f.p as p
+```
+
+The forwarded expression and the export target must both be static expressions,
+and like the export statement do not participate in last-connect semantics.
+
+Forward statements may narrow a probe of an aggregate to a sub-element using
+static expression:
+
+```firrtl
+module Foo :
+  output p : Probe<UInt[2]>[2]
+  ; ...
+
+module Forward :
+  output p : Probe<UInt>
+
+  inst f of Foo
+  forward f.p[0][1] as p
```

Comment:
@azidar does this go for the other statements like `export foo as p` as well?

Reply:
What's a language spec PR for if not to discuss syntax? :wink: .  In all seriousness, thanks for mentioning this and I hope to find at least somewhat better syntax for this and other concerns.

I'd like to find a better syntax here, see other discussion re:possibly combining forward and export into something more general (although don't know want to introduce it in a fully general way just yet regardless, reference-only for now, "one thing at a time").

FWIW the direction change is specifically intended to ensure there's no confusing this with a "standard" connection, as it's not a connection at all in the sense FIRRTL uses the notion and maybe helps w/the graphical intuition given re:last-connects and when statements.  This might not be as useful/effective as I hoped.

Previously was using a right arrow as part of that notation to unify w/connections today (kinda like a connection "BUT DIFFERENT"), which hopefully makes it a little less confusing (`x => y` surely doesn't describe `y` going to `x`, if that's read as an arrow anyway), but that syntax is taken by parameters on memories + registers and weakly suggests this is maybe something like a reverse connection.  But don't want to take things off the table unnecessarily :+1: .

I don't love the `forward y from x` I find that incredibly confusing and if oriented that way I vote it's no longer called "forward" as that just reads poorly (to me/IMHO).

I've been playing around with alternative terminology, how about `define y = x`  (if we like it the credit belongs to someone else to be named later :wink:)?  Idea for that language is to clarify the static nature and that it's not reasonable to define the same thing multiple times / differently, at least to me.  "assign" was my preference previously but especially since that's already a Verilog thing (that doesn't mean what we want) am hoping to find something else.

Reply:
Yeah I like `define`

Reply:
Unified export and forward into `define`, which is right-to-left.  Thanks!